### PR TITLE
Get some of the limited API working with Py_LIMITED_API defined

### PR DESCRIPTION
**Pull request by da-woods**
This gets a minimal limited API example working with Py_LIMITED_API defined, and adds a small test to prove it.

### CODE REVIEW

File: Cython/Utility/FunctionArguments.c
Comment (scoder) [inline Cython/Utility/FunctionArguments.c:246]: The `GET_ITEM` call should rather be guarded by `SAFE_MACROS` and `BORROWED_REFS`.

File: Cython/Utility/FunctionArguments.c
Comment (scoder) [inline Cython/Utility/FunctionArguments.c:417]: Same here, we shouldn't unconditionally use the macro.

File: Cython/Utility/ObjectHandling.c
Comment (scoder) [inline Cython/Utility/ObjectHandling.c:2941]: Isn't there a `NULL` end-marker missing for the arguments?

File: Cython/Utility/TypeConversion.c
Comment (scoder) [inline Cython/Utility/TypeConversion.c:832]: Ok, that looks expensive. But having it work at all is probably still better than having it fail..

Comment (da-woods) [reply Cython/Utility/TypeConversion.c:832]: This is the fallback code for "int type is bigger than long long" or "int type is bigger than long and Python doesn't define `PY_LONG_LONG`". So practically I doubt it gets used much.

But yeah, I expect it's expensive.

File: Cython/Utility/FunctionArguments.c
Comment (da-woods) [reply Cython/Utility/FunctionArguments.c:246]: Still thinking about this. In both cases the `BORROWED_REFS` guard looks hard to implement, and isn't necessary for limited API support (where we allow borrowed refs by default).

So I'd rather just do `SAFE_MACROS` at this stage.

I've moved `__Pyx_PyTuple_GET_SIZE` inside `SAFE_MACROS` but I don't think this is the right solution so may come back and change it.

Comment (da-woods) [reply Cython/Utility/FunctionArguments.c]: I think these should now be handled (at the cost of some complication)

**Comment by da-woods** [comment]
Closing and re-opening (there's an extra commit that github doesn't seem to have noticed...)

File: Cython/Utility/Exceptions.c
Comment (scoder) [inline Cython/Utility/Exceptions.c:964]:
```suggestion
        // limited API. The check against None should cover the most
```

File: Cython/Utility/Exceptions.c
Comment (scoder) [inline Cython/Utility/Exceptions.c:973]: I wonder if we should really look at `Py_VERSION_HEX` here or the requested Limited API version. (And I wonder if that doesn't generally apply in all Limited API sections…)

File: Cython/Utility/Exceptions.c
Comment (scoder) [inline Cython/Utility/Exceptions.c:920]: I'll just ignore this part in my review and believe both properties as you state them here.

File: Cython/Utility/FunctionArguments.c
Comment (scoder) [inline Cython/Utility/FunctionArguments.c:263]: Does this look any better if you reverse the conditions to put `#if AVOID_BORROWED_REFS` first and `#if SAFE_MACROS` second?

File: Cython/Utility/FunctionArguments.c
Comment (scoder) [inline Cython/Utility/FunctionArguments.c:276]: Isn't this guard the wrong way round? And, should we really keep the two borrowed references around for so long below? We might want to incref both any way, don't you think? At least if the fast pointer comparison below fails and we need to do actual comparisons. I don't think those are always safe while holding on to borrowed references.

File: Cython/Utility/FunctionArguments.c
Comment (scoder) [inline Cython/Utility/FunctionArguments.c:466]: Seems worth reversing the two cases.

File: Cython/Utility/ModuleSetupCode.c
Comment (scoder) [inline Cython/Utility/ModuleSetupCode.c:658]: Weren't macros with `...` arguments an issue in C89? I faintly recall that this was a gcc extension at the time.

File: Cython/Utility/FunctionArguments.c
Comment (da-woods) [reply Cython/Utility/FunctionArguments.c]: The guard was definitely the wrong way round. I've changed it so key and value are universally reference counted in this function to avoid having to think too much about what's safe.

File: Cython/Utility/Exceptions.c
Comment (da-woods) [reply Cython/Utility/Exceptions.c]: In this particular case `(struct _frame*) looks to work on all Python versions so I've removed the condition.

In general you're right - we usually want to use the limited API version in Limited API sections (I've exposed that with a `__PYX_LIMITED_VERSION_HEX` which will be `PY_VERSION_HEX` in non-limited API mode). Here we did want to use `PY_VERSION_HEX` but that's probably a defect in CPython really.

**Comment by astrofrog** [comment]
@da-woods - thank you for working on this and getting my example working, this is very exciting!

Just to check, what would be most helpful if I run into issues with more complex examples, should I open issues with minimal failing examples, or contribute an xfailing regression test? (or something else?)

**Comment by da-woods** [comment]
> @da-woods - thank you for working on this and getting my example working, this is very exciting!
> 
> Just to check, what would be most helpful if I run into issues with more complex examples, should I open issues with minimal failing examples, or contribute an xfailing regression test? (or something else?)

@astrofrog I have one follow-up to this PR that gets a minimal cdef class working. The work is done on that but I'll submit it separately just to keep the size of this down a bit. So wait for that!

After that, opening issues is probably the most useful.

The long-term plan would be to run as much of the main Cython test-suite as possible with the limited API rather than have unique dedicated limited API tests. But it's not completely obvious what the best way to get there is yet

**Comment by aws-taylor** [comment]
Great work! One more small one - https://github.com/cython/cython/blame/6ad6ca0e9e7d030354b7fe7d7b56c3f6e6a4bc23/Cython/Utility/Exceptions.c#L19 - Looks like `Py_OptimizeFlag` is getting used, but it doesn't appear to be part of the limited API.

**Comment by da-woods** [comment]
> Great work! One more small one - https://github.com/cython/cython/blame/6ad6ca0e9e7d030354b7fe7d7b56c3f6e6a4bc23/Cython/Utility/Exceptions.c#L19 - Looks like `Py_OptimizeFlag` is getting used, but it doesn't appear to be part of the limited API.

Thanks for the report. I've created a new issue - just want to keep this PR to the current scope. It's likely a fairly small change but it's one of many fairly small changes

File: Cython/Utility/TypeConversion.c
Comment (scoder) [reply Cython/Utility/TypeConversion.c:832]: BTW, `PY_LONG_LONG` is now always defined in recent CPython versions. I think since back when they switched to C99 or a bit later, when they noticed that their code actually doesn't compile without `HAVE_LONG_LONG` any more. :)

File: Cython/Utility/FunctionArguments.c
Comment (scoder) [inline Cython/Utility/FunctionArguments.c:465]: Let's keep the order the same as for the decref macros below. We could even merge the two blocks since the definitions belong together, really.
```suggestion
#if CYTHON_AVOID_BORROWED_REFS
    // This is the only case where we request an owned reference.
    #define __Pyx_Arg_VARARGS(args, i) PySequence_GetItem(args, i)
#elif CYTHON_ASSUME_SAFE_MACROS
    #define __Pyx_Arg_VARARGS(args, i) PyTuple_GET_ITEM(args, i)
#else
    #define __Pyx_Arg_VARARGS(args, i) PyTuple_GetItem(args, i)
#endif
```

File: Cython/Utility/ObjectHandling.c
Comment (scoder) [inline Cython/Utility/ObjectHandling.c:2947]: This also looks really slow, given that we'd do this on each method lookup. It's ok to get it working this way, but I doubt that it can stay that way.

In general, we should try not to be visibly slower than CPython, at least on stuff that happens repeatedly. We can assume that users of the Limited API do not care as much about performance as others, but they'd probably still not want their code to run much slower than in plain Python.

**Review by scoder — APPROVED** [review]
This makes many things work that didn't work at all before, and I think it's in a good shape to merge now. We'll see how the performance is, but in any case, this isn't the last thing we'll do for the Limited API, and optimisations can come later. We can also play some of the missing bits back to CPython and discuss better solutions with them.

File: Cython/Utility/Exceptions.c
Comment (scoder) [inline Cython/Utility/Exceptions.c:938]: The `.replace()` method seems to require Py3.8+ and is missing in Py3.7, it seems.

File: Cython/Utility/ObjectHandling.c
Comment (da-woods) [reply Cython/Utility/ObjectHandling.c:2947]: It might be worth double checking if we actually use this. I think it's mainly used in CythonFunction.c, which has been rewritten substantially for the limited API anyway, so it's possible that we can just drop this.

File: Cython/Utility/Exceptions.c
Comment (da-woods) [reply Cython/Utility/Exceptions.c:961]: Yeah - it'd noticed that it could be cached, but was just trying to keep things simple for the initial version.

We can definitely come back and optimize this later.

I'm also open to different approaches, but I honestly couldn't think of anything else that might work.

File: Cython/Utility/ModuleSetupCode.c
Comment (da-woods) [inline Cython/Utility/ModuleSetupCode.c:655]: I've also refactored this slightly again, since it used `replace` and so was also broken on Py3.7. Might be a bit more performant than before, but probably relies on somewhat unstable (Python) API

**Comment by aws-taylor** [comment]
Two other small ones related to the (non) existence of `tp_free` in PyTypeObject:

* https://github.com/da-woods/cython/blob/limited-api-with-py-limited-api/Cython/Compiler/ModuleNode.py#L1826C13-L1826C53
* https://github.com/da-woods/cython/blob/limited-api-with-py-limited-api/Cython/Compiler/ModuleNode.py#L3403

`code.putln("(*Py_TYPE(o)->tp_free)(o);")`

Reading through https://hg.python.org/cpython/rev/655d7a55c165, it seems like the fix is probably something along the lines of:

`code.putln("((freefunc)PyType_GetSlot(Py_TYPE(o), Py_tp_free))(o);")`

**Comment by da-woods** [comment]
> Two other small ones related to the (non) existence of `tp_free` in PyTypeObject:

The follow up PR to this will do classes. I'd got the first of those lines. I hadn't got the second, but I'll make sure it add it - thanks

**Comment by aws-taylor** [comment]
> > Two other small ones related to the (non) existence of `tp_free` in PyTypeObject:
> 
> The follow up PR to this will do classes. I'd got the first of those lines. I hadn't got the second, but I'll make sure it add it - thanks

Awesome. Thanks again for this change; I'm really hoping this can get out the door. If it wasn't obvious, I'm working through a stack of errors for my own codebase. Let me know if you'd prefer I open tickets as I come across issues or just piggy-back here.

**Comment by aws-taylor** [comment]
Another 2 I came across related to tp_basicsize:
* https://github.com/cython/cython/blob/db995de3d24fe463a4121ab9acb94c968d4f25b3/Cython/Compiler/ModuleNode.py#L1582
* https://github.com/cython/cython/blob/db995de3d24fe463a4121ab9acb94c968d4f25b3/Cython/Compiler/ModuleNode.py#L1818

Unlike tp_free, there doesn't appear to be a slot macro for `Py_tp_basicsize`. I think this is because under the limited API, there are no built-in types, and thus there are no special types for which sizeof != tp_basicsize. If that's true, then I think the fix may be to simply ignore this clause in the `if` statement when compiling with the limited API, because it'll always be true. There may also be a fix involving `basicsize_builtins_map`.

**Comment by da-woods** [comment]
The PyPy failure looks to be a mess up with fused function optional arguments.

The code it's generating is

```
values[3] = __Pyx_Arg_NEWREF(((PyObject *)((long)((long)2))));
```

With the `__Pyx_Arg_NEWREF` being added by this PR, and being the bit that's breaking. However it's plainly wrong even without this PR - it just isn't generating invalid reference counting.

Unfortunately this pretty much blocks this PR until that's sorted I think.

**Comment by aws-taylor** [comment]
Another 1-liner related to `tp_dictoffset`:

* https://github.com/cython/cython/blob/c0870f912196a89a6bc24f94f2684f2633d75589/Cython/Utility/ExtensionTypes.c#L138

Per https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_dictoffset, `tp_dictoffset` is supposed to be considered write-only.  I think the fix might be:

`if (dictoffset == 0 && PyObject_GenericGetDict(b, NULL) )`

**Comment by da-woods** [comment]
> Another 1-liner related to `tp_dictoffset`:

This is already fixed in the follow-up

> tp_basicsize

This isn't. It probably isn't too hard - it's already conditional on `CYTHON_COMPILING_IN_CPYTHON` (which is off with the limited API) - the condition just needs to be moved to exclude the code, rather than being part of the if-statement

**Comment by da-woods** [comment]
Depends on https://github.com/cython/cython/pull/5614

File: Cython/Utility/ModuleSetupCode.c
Comment (da-woods) [inline Cython/Utility/ModuleSetupCode.c:739]: I think these would be better read from `inspect` when the module is initialize (i.e. much like the debug flag) - it feels a bit against the spirit of the limited API to just redefine constants that have been intentionally excluded.

With that said, I'll create an issue when this PR is done rather than try to make too many more changes.

**Pull request merged**
